src/engine.rs

```diff
@@ -218,31 +220,38 @@ impl Reedline {
         for command in commands {
             match command {
                 EditCommand::InsertChar(c) => {
-                    let search = self
-                        .history_search
-                        .as_mut()
-                        .expect("couldn't get history_search as mutable"); // We checked it is some
-                    search.step(BasicSearchCommand::InsertChar(*c), &self.history);
+                    let navigation = self.history.get_navigation();
+                    if let HistoryNavigationQuery::SubstringSearch(substring) = navigation {
+                        let new_string = format!("{}{}", substring, c);
+                        self.history
+                            .set_navigation(HistoryNavigationQuery::SubstringSearch(new_string));
+                    } else {
+                        self.history
+                            .set_navigation(HistoryNavigationQuery::SubstringSearch(format!(
+                                "{}",
+                                c
+                            )))
+                    }
```

Comment:
There should be a better way to append string here.

Reply:
If the History trait had some way of adding to the history, that seems like it would simplify the API call there

Reply:
One idea would be for the history call to take a callback. In the callback, it could pass you what the current string is, and then from there you could append or change the string.

That said, I'm all for quick-and-dirty first, and then going back over it and cleaning it up once it works.

Reply:
I will revist this point again in the future if that is acceptable.